electron/src/lib/AutomatedSingleSignOn.ts

```diff
@@ -0,0 +1,148 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see http://www.gnu.org/licenses/.
+ *
+ */
+
+import {WebContents, WebviewTag, app, clipboard} from 'electron';
+import {BaseError} from 'make-error-cause';
+import {URL} from 'url';
+const dialog = require('electron').dialog || require('electron').remote.dialog;
+import * as windowManager from '../js/window-manager';
+import {getText} from '../locale/locale';
+import {EVENT_TYPE} from './eventType';
+
+class MaximumAccountsReachedError extends BaseError {}
+
+class AutomatedSingleSignOn {
+  private static readonly SSO_ACTION_DELAY = 1000;
+  private static readonly SSO_LOGIN_HASH = 'login';
+  private static readonly SSO_CODE_MAX_LENGTH = 255;
+  public static readonly clipboard = {
+    read: () => clipboard.readText(),
+    write: (value: string) => clipboard.writeText(value),
+  };
+  private static readonly sleep = (ms: number) => new Promise(resolve => setTimeout(resolve, ms));
+  public static readonly handleProtocolRequest = async (route: URL) => {
+    if (typeof route.pathname !== 'string') {
+      return;
+    }
+
+    const sendCodeToRenderer = () => {
+      const code = route.pathname.trim().substr(1);
+      main.webContents.send(EVENT_TYPE.ACCOUNT.SSO_LOGIN, code);
+    };
+
+    await app.whenReady();
+
+    const main = windowManager.getPrimaryWindow();
+    if (main.webContents.isLoading()) {
+      // App is booting
+      main.webContents.once('did-finish-load', () => sendCodeToRenderer());
+    } else {
+      if (!main.isVisible()) {
+        main.show();
+        main.focus();
+      }
+      sendCodeToRenderer();
+    }
+  };
+  private oldClipboard?: string;
+  private webview!: WebviewTag;
+  private webContents!: WebContents;
+
+  constructor(private readonly ssoCode: string) {}
+
+  private buildUrl() {
+    if (!this.webview.src) {
+      return 'about:blank';
+    }
+    const src = new URL(this.webview.src);
+    src.pathname = '/auth/';
+    src.hash = AutomatedSingleSignOn.SSO_LOGIN_HASH;
+    return src.href;
+  }
+
+  private async clickOn(element: string) {
+    await this.webContents.executeJavaScript(`document.querySelector('${element}').click();`);
+    await AutomatedSingleSignOn.sleep(AutomatedSingleSignOn.SSO_ACTION_DELAY);
+  }
+
+  private executeLogin() {
+    return new Promise((resolve, reject) => {
+      const failedToLoad = () => reject('Webview failed to load');
+      this.webview = document.querySelector('.Webview:not(.hide)') as WebviewTag;
+      this.webContents = this.webview.getWebContents();
+
+      this.webContents.once('did-fail-load', failedToLoad);
+      this.webContents.once('did-finish-load', async () => {
+        this.webContents.removeListener('did-fail-load', failedToLoad);
+        await this.clickOn('a[data-uie-name="go-sign-in-sso"]');
```

Comment:
Can we directly go to the link instead of going to a link via a simulated click?

Reply:
We can't, Electron have an issue where the clipboard cannot be read *every time* if we directly go load the #sso page.

Reply:
When using hash URLs on webapp then it can log you in automatically (if the code is correct):
- https://github.com/wireapp/wire-webapp/pull/5753

Do we still need `this.clickOn('a[data-uie-name="go-sign-in-sso"]')` when the webapp logs you in automatically?

Reply:
This has been removed.